### ISSUE TITLE
AMP Query Not Displaying In WP_Query on Posts [Won't Validate]

### ISSUE THREAD

**Issue by jasonbiondo** [ampproject/amp-wp#2354]
I have been troubleshooting this issue with the AMP install. For some reason when doing a var_dump on $wp_query amp is missing and causing the pages not to validate.  

Any idea why this might be happening?  The main homepage of the blog does have amp in wp_query. This function is in the helpers file
 
![image](https://user-images.githubusercontent.com/1186147/58044848-cc5e4700-7b0e-11e9-99ff-3e667022466c.png)

![image](https://user-images.githubusercontent.com/1186147/58045908-db92c400-7b11-11e9-8e2b-08352f17770f.png)

**Comment by westonruter** [comment]
That would be expected, as the output would be getting sent before the `<html>`. Instead of doing `var_dump()` I suggest tailing the PHP error log and looking at the output there.

Otherwise, you could defer printing until `wp_footer` action,  for example:

```
add_action( 'wp_footer', function() use ( $wp_query ) {
    var_dump( $wp_query );
} );
```

But yeah, I recommend the error log approach. I usually do this:

```php
error_log( json_encode( $wp_query, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES ) );
```

The use of `json_encode()` ensures the output will work even when the sanitizers are running (since `var_dump()` uses output buffering under the hood, IIRC).

**Comment by jasonbiondo** [comment]
@westonruter The error I get back is:

while reading response header from upstream, client: 13.59.13.249, server: dev.mysite.com, request: "GET /travel-blog/2019/05/20/testing/?amp&amp_validate=7e8c97e72f&amp_cache_bust=3656782425 HTTP/1.1", upstream: "fastcgi://unix:/var/run/php/php7.2-fpm.sock:", host: "dev.mysite.com", referrer: "http://dev.mysite.com/travel-blog/2019/05/20/testing/?amp&amp_validate=7e8c97e72f&amp_cache_bust=3656782425"
2019/05/20 15:26:41 [error] 11730#11730: *14128 FastCGI sent in stderr: "PHP message: {

Any idea why?

**Comment by westonruter** [comment]
@jasonbiondo are you using the latest version of the AMP plugin? Line 293 has an `isset()` check:

https://github.com/ampproject/amp-wp/blob/885e4888a284e5c703779b5af89620358cab90db/includes/amp-helper-functions.php#L292-L300

This would prevent any such PHP notice.

**Comment by jasonbiondo** [comment]
@westonruter Yes, I am using the latest version, but took your advice and added in:

![image](https://user-images.githubusercontent.com/1186147/58055273-38e73f00-7b2b-11e9-90ab-137e0ebc0911.png)

**Comment by westonruter** [comment]
You may be calling `is_amp_endpoint()` too early. Are you calling it after the `parse_query` action? And actually it needs to be done at/after the `wp` action.

**Comment by jasonbiondo** [comment]
I have made no changes to the plugin code and I am using a fresh install of wordpress.  How can I verify this?

**Comment by westonruter** [comment]
I guess I don't understand where the original problem is coming from. What theme and plugins are you using?

**Comment by westonruter** [comment]
> wp_query is not returning the amp as a parameter in the array even though ?amp is in the url. In fact no query string is returned on a blog post regardless of what I set ?somevariable to.

Even so this should not result in an “undefined index” notice. The `\WP_Query::get()` method also includes an `isset()` check.

**Comment by jasonbiondo** [comment]
The main issue is when I go to enable either native or transitional I get an error: "Transitional mode activated! However, there was an error when checking the AMP validity for your site. URL validation failed to due to the absence of the expected JSON-containing AMP_VALIDATION comment after the body." 

Even with wp_debug enabled there are no errors in the source code and no errors in the logs.  When I try to validate a blog post I get brought to the plugin page with the amp validation error and I have no idea why.  

This video shows the error: https://youtu.be/vqNdLCcYb_Q

**Comment by westonruter** [comment]
@jasonbiondo Thanks for that. Often that indicates that there was a fatal error during the post-processor phase. Can you look at your PHP error log for any such errors?

**Comment by jasonbiondo** [comment]
Thats actually the main problem.   There is nothing in the error log indicating a problem.

**Comment by westonruter** [comment]
@jasonbiondo Can you DM me on Twitter or WordPress Slack to debug further?

**Comment by jasonbiondo** [comment]
@westonruter I tweeted you my email to invite me on slack

**Comment by jasonbiondo** [comment]
For those that might hit this problem... There is a breaking changing in the latest wordpress build causing amp to not be in wp_query.  Use an older build and the plugin works just fine.

**Comment by westonruter** [comment]
So it is a WordPress core issue? Introduced in 5.2? Is it fixed in 5.2.1?

**Comment by jasonbiondo** [comment]
It is a wordpress core issue in 5.2.  I don't know if it's fixed in 5.2.1.  The issue is not present in 5.1.

**Comment by westonruter** [comment]
WordPress 5.2.1 is now available: https://wordpress.org/news/2019/05/wordpress-5-2-1-maintenance-release/
